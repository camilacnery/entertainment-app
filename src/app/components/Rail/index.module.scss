.rail {
  color: #FFFFFF;
  padding: 0 0 6.5vh 4.7vw;
}

.railList {
  display: flex;
  align-items: center;
  overflow-x: scroll;
  overflow: hidden;
  padding: 3.3vh 0;
  height: 38.5vh;
  position: relative;
}

.railListItem {
  position: relative;
  display: flex;
  flex-direction: column;
}
