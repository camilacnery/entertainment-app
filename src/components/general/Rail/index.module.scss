@use '@/styles/constants' as *;

.rail {
  padding: 0 0 0 2em ;
  width: 100vw;
}

.railList {
  display: flex;
  gap: 1em;
  overflow-x: scroll;
  overflow-y: hidden;
  padding: 1em;

  &::-webkit-scrollbar {
    display: none;
  }
}

.railListItem {
  position: relative;
  cursor: pointer;
  transition: transform 450ms;
  border-radius: $border-radius;
  background-color: $background-box-color;
  min-width: 200px;
  min-height: 300px;
  
  img {
    z-index: 3;
    position: relative;
    border-radius: 5px;
    max-height: 300px;
    min-height: 100%;
  }

  &:hover {
    transform: scale(1.1);
  }
}

.railListItemTitle {
  z-index: 1;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  text-align: center;
  font-size: 1.6rem;
}
