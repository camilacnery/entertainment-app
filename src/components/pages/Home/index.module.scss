.main {
  min-height: 100vh;
  padding: 8rem 0 4rem 0;
  flex: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.rails {
  z-index: 2;
}
