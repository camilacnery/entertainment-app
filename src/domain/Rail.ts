export interface IRailItem {
  posterUrl: string;
  title: string;
  id: number;
}

export interface IRail {
  name: string;
  items: IRailItem[];
}
