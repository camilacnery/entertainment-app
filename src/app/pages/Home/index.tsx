import Head from "next/head";
import { FC } from "react";
import { TRail } from "../../../domain/Rail";
import Rail from "../../components/Rail";
import styles from "./index.module.scss";

type TProps = {
  homeRails: TRail[];
};

const HomePage: FC<TProps> = ({ homeRails }) => {
  return (
    <div className={styles.container}>
      <Head>
        <title>Create Next App</title>
        <meta name="description" content="Generated by create next app" />
        <link rel="icon" href="/favicon.ico" />
      </Head>

      <main className={styles.main}>
        <h1 className={styles.title}>Movie Catalog</h1>

        {homeRails.map((rail) => (
          <Rail key={`rail-${rail.name}`} {...rail} />
        ))}
      </main>

      <footer className={styles.footer}></footer>
    </div>
  );
};

export default HomePage;
