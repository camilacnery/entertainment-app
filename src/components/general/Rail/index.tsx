import { FC } from "react";
import Image from "next/future/image";
import { TRail } from "../../../domain/Rail";
import styles from "./index.module.scss";

const Rail: FC<TRail> = ({ name, items }) => {
  return (
    <section className={styles.rail}>
      <h1>{name}</h1>
      <div className={styles.railList}>
        {items.map((item) => (
          <div key={`rail-item-${item.title}`} className={styles.railListItem}>
            <Image
              loading="lazy"
              width={200}
              height={280}
              src={item.posterUrl}
              alt={`Movie poster for ${item.title}`}
            />
          </div>
        ))}
      </div>
    </section>
  );
};

export default Rail;
