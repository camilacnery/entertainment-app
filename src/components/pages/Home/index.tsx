import { FC } from "react";
import { IRail } from "@/domain/Rail";
import Rail from "@/components/general/Rail";
import Layout from "@/components/general/Layout";

type TProps = {
  homeRails: IRail[];
};

const HomePage: FC<TProps> = ({ homeRails }) => {
  const backgroundUrl = homeRails[0]?.items[0]?.backdropUrl;

  return (
    <Layout
      meta={{ title: "The Catalog", description: "List of movies, movie details, and more" }}
      backgroundUrl={backgroundUrl}
    >
      {homeRails.map((rail) => (
        <Rail key={`rail-${rail.name}`} {...rail} />
      ))}
    </Layout>
  );
};

export default HomePage;
