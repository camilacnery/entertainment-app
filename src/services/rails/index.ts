import { IRail, IRailItem } from "@/domain/Rail";
import tmdbClient, { ITMDBList } from "@/clients/tmdb";

const buildRailItem = (result: ITMDBList["results"][0]) => ({
  id: result.id,
  posterUrl: tmdbClient.buildImageUrl(result.poster_path),
  backdropUrl: tmdbClient.buildImageUrl(result.backdrop_path),
  title: result.name || result.title,
});

const getHomeRails = async (): Promise<IRail[]> => {
  const rails = [
    { name: "Trending", path: "/trending/movie/week" },
    { name: "Top Rated", path: "/movie/top_rated" },
    { name: "Action", path: "/discover/movie", filter: "with_genres=28" },
    { name: "Comedy", path: "/discover/movie", filter: "with_genres=35" },
    { name: "Drama", path: "/discover/movie", filter: "with_genres=18" },
  ];

  const homeRails: IRail[] = [];

  for (const rail of rails) {
    try {
      const contentList = await tmdbClient.request<ITMDBList>(rail.path, rail.filter);

      if (contentList.results?.length) {
        homeRails.push({
          name: rail.name,
          items: contentList.results.map(itemMapper),
        });
      }
    } catch {}
  }

  return homeRails;
};

const getMovieRecommendationRail = async (movieId: string): Promise<IRailItem[]> => {
  const rail = await tmdbClient.request<ITMDBList>(`/movie/${movieId}/recommendations`);

  return rail.results?.map(itemMapper) || [];
};

export { getHomeRails, getMovieRecommendationRail };
