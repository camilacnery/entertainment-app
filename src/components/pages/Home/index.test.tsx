import { screen, render, within } from "@testing-library/react";
import { IRail } from "@/domain/Rail";
import { buildRail } from "@/tests/builders/rails";
import Home from ".";

const pageElements = {
  rail: {
    list: (name: string) => screen.queryByRole("list", { name }),
    items: (rail: HTMLElement) => within(rail).getAllByRole("listitem"),
    movie: (rail: HTMLElement, movieName: string) =>
      within(rail).queryByRole("img", {
        name: movieName,
      }),
  },
};

const renderHome = (rails: IRail[]) => {
  return render(<Home homeRails={rails} />);
};

describe("Home", () => {
  it("renders home with multiple rails", () => {
    const rails = [
      buildRail({ name: "Trending", size: 2 }),
      buildRail({ name: "Top Rated", size: 3 }),
    ];

    const { baseElement } = renderHome(rails);

    expect(baseElement).toMatchSnapshot();

    // eslint-disable-next-line  @typescript-eslint/no-non-null-assertion
    const trendingRail = pageElements.rail.list("Trending")!;
    expect(pageElements.rail.items(trendingRail)).toHaveLength(2);
    expect(pageElements.rail.movie(trendingRail, "First Movie")).toBeInTheDocument();
    expect(pageElements.rail.movie(trendingRail, "Second Movie")).toBeInTheDocument();

    // eslint-disable-next-line  @typescript-eslint/no-non-null-assertion
    const topRatedRail = pageElements.rail.list("Top Rated")!;
    expect(pageElements.rail.items(topRatedRail)).toHaveLength(3);
    expect(pageElements.rail.movie(topRatedRail, "First Movie")).toBeInTheDocument();
    expect(pageElements.rail.movie(topRatedRail, "Second Movie")).toBeInTheDocument();
    expect(pageElements.rail.movie(topRatedRail, "Third Movie")).toBeInTheDocument();
  });

  it("does not render rails with empty items", () => {
    const rails = [
      buildRail({ name: "Trending", size: 0 }),
      buildRail({ name: "Top Rated", size: 3 }),
    ];

    renderHome(rails);

    expect(pageElements.rail.list("Trending")).not.toBeInTheDocument();
    expect(pageElements.rail.list("Top Rated")).toBeInTheDocument();
  });
});
