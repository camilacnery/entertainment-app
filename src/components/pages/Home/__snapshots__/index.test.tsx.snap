// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders home with multiple rails 1`] = `
<body>
  <div>
    <div
      class="banner"
    >
      <img
        alt=""
        data-nimg="future"
        decoding="async"
        height="768"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2Fzb6fM1CX41D9rF9hdgclu0peUmy.jpg&w=2048&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2Fzb6fM1CX41D9rF9hdgclu0peUmy.jpg&w=1080&q=75 1x, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2Fzb6fM1CX41D9rF9hdgclu0peUmy.jpg&w=2048&q=75 2x"
        style="color: transparent;"
        width="1024"
      />
    </div>
    <main
      class="main"
    >
      <div
        class="content"
      >
        <section
          class="rail"
        >
          <h1
            id="rail-heading-Trending"
          >
            Trending
          </h1>
          <ul
            aria-labelledby="rail-heading-Trending"
            class="railList"
          >
            <li
              class="railListItem"
            >
              <img
                alt="Schindler's List"
                data-nimg="future"
                decoding="async"
                height="280"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FsF1U4EUQS8YHUYjNl3pMGNIQyr0.jpg&w=640&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FsF1U4EUQS8YHUYjNl3pMGNIQyr0.jpg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FsF1U4EUQS8YHUYjNl3pMGNIQyr0.jpg&w=640&q=75 2x"
                style="color: transparent;"
                width="200"
              />
            </li>
            <li
              class="railListItem"
            >
              <img
                alt="The Godfather"
                data-nimg="future"
                decoding="async"
                height="280"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2F3bhkrj58Vtu7enYsRolD1fZdja1.jpg&w=640&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2F3bhkrj58Vtu7enYsRolD1fZdja1.jpg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2F3bhkrj58Vtu7enYsRolD1fZdja1.jpg&w=640&q=75 2x"
                style="color: transparent;"
                width="200"
              />
            </li>
          </ul>
        </section>
        <section
          class="rail"
        >
          <h1
            id="rail-heading-TopRated"
          >
            Top Rated
          </h1>
          <ul
            aria-labelledby="rail-heading-TopRated"
            class="railList"
          >
            <li
              class="railListItem"
            >
              <img
                alt="Schindler's List"
                data-nimg="future"
                decoding="async"
                height="280"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FsF1U4EUQS8YHUYjNl3pMGNIQyr0.jpg&w=640&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FsF1U4EUQS8YHUYjNl3pMGNIQyr0.jpg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FsF1U4EUQS8YHUYjNl3pMGNIQyr0.jpg&w=640&q=75 2x"
                style="color: transparent;"
                width="200"
              />
            </li>
            <li
              class="railListItem"
            >
              <img
                alt="The Godfather"
                data-nimg="future"
                decoding="async"
                height="280"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2F3bhkrj58Vtu7enYsRolD1fZdja1.jpg&w=640&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2F3bhkrj58Vtu7enYsRolD1fZdja1.jpg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2F3bhkrj58Vtu7enYsRolD1fZdja1.jpg&w=640&q=75 2x"
                style="color: transparent;"
                width="200"
              />
            </li>
            <li
              class="railListItem"
            >
              <img
                alt="Pulp Fiction"
                data-nimg="future"
                decoding="async"
                height="280"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FfIE3lAGcZDV1G6XM5KmuWnNsPp1.jpg&w=640&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FfIE3lAGcZDV1G6XM5KmuWnNsPp1.jpg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fimage.tmdb.org%2Ft%2Fp%2Foriginal%2FfIE3lAGcZDV1G6XM5KmuWnNsPp1.jpg&w=640&q=75 2x"
                style="color: transparent;"
                width="200"
              />
            </li>
          </ul>
        </section>
      </div>
    </main>
  </div>
</body>
`;
